File: malai-core/org.malai.jinput/src/main/org/malai/jinput/interaction/library/ButtonPressed.java
package org.malai.jinput.interaction.library;

import net.java.games.input.Component;

import org.malai.interaction.TerminalState;
import org.malai.jinput.interaction.ButtonPressedTransition;
import org.malai.swing.interaction.SwingInteraction;

/**
 * A ButtonPressed interaction occurs when a button is pressed.<br>
 * <br>
 * This file is part of Malai.<br>
 * Copyright (c) 2009-2013 Arnaud BLOUIN<br>
 * <br>
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 05/19/2010<br>
 * @author Arnaud BLOUIN
 * @since 0.1
 */
public class ButtonPressed extends SwingInteraction {
	/** The pressed button. */
	protected Component.Identifier.Button button;

	/**
	 * Creates the interaction.
	 */
	public ButtonPressed() {
		super();
		initStateMachine();
	}


	@Override
	public void reinit() {
		super.reinit();

		button = null;
	}


	@SuppressWarnings("unused")
	@Override
	protected void initStateMachine() {
		final TerminalState pressed = new TerminalState("pressed"); //$NON-NLS-1$

		addState(pressed);

		new ButtonPressedTransition(initState, pressed) {
			@Override
			public void action() {
				super.action();

				ButtonPressed.this.button = this.button;
			}
		};
	}


	/**
	 * @return The pressed button.
	 * @since 0.1
	 */
	public Component.Identifier.Button getButton() {
		return button;
	}
}

File: malai-core/org.malai.ts/src/interaction/library/ButtonPressed.ts
/*
 * This file is part of Malai.
 * Copyright (c) 2009-2018 Arnaud BLOUIN
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

import {FSMDataHandler} from "../FSMDataHandler";
import {TSFSM} from "../TSFSM";
import {ButtonPressedTransition} from "../ButtonPressedTransition";
import {TerminalState} from "../../../src-core/fsm/TerminalState";
import {isButton} from "../Events";
import {TSInteraction} from "../TSInteraction";


class ButtonPressedFSM extends TSFSM<ButtonPressedFSMHandler> {
    public constructor() {
        super();
    }

    public buildFSM(dataHandler?: ButtonPressedFSMHandler): void {
        if (this.states.length > 1) {
            return;
        }
        super.buildFSM(dataHandler);
        const pressed: TerminalState<Event> = new TerminalState<Event>(this, "pressed");
        this.addState(pressed);

        new class extends ButtonPressedTransition {
            public action(event: Event): void {
                if (event.target !== null && isButton(event.target) && dataHandler !== undefined) {
                    dataHandler.initToPressedHandler(event);
                }
            }
        }(this.initState, pressed);
    }
}

interface ButtonPressedFSMHandler extends FSMDataHandler {
    initToPressedHandler(event: Event): void;
}

/**
 * A user interaction for buttons.
 * @author Arnaud BLOUIN
 */
export class ButtonPressed extends TSInteraction<ButtonPressedFSM, Element> {
    private readonly handler: ButtonPressedFSMHandler;

    /**
     * Creates the interaction.
     */
    public constructor() {
        super(new ButtonPressedFSM());

        this.handler = new class implements ButtonPressedFSMHandler {
            private readonly _parent: ButtonPressed;

            constructor(parent: ButtonPressed) {
                this._parent = parent;
            }

            public initToPressedHandler(event: Event): void {
                if (event.target !== null && isButton(event.target)) {
                    this._parent._widget = event.currentTarget as Element;
                }
            }

            public reinitData(): void {
                this._parent.reinitData();
            }
        }(this);

        this.fsm.buildFSM(this.handler);
    }

    public onNewNodeRegistered(node: EventTarget): void {
        if (isButton(node)) {
            this.registerActionHandler(node);
        }
    }

    public onNodeUnregistered(node: EventTarget): void {
        if (isButton(node)) {
            this.unregisterActionHandler(node);
        }
    }
}
